tuxemon/lib/compress_json.py

```diff
@@ -0,0 +1,214 @@
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+"""
+import json
+from typing import Dict, Any
+import traceback
+import os
+
+__all__ = [
+    "dump",
+    "load",
+]
+
+_DEFAULT_EXTENSION_MAP = {
+    "json": "json",
+    "gz": "gzip",
+    "save": "gzip",
+    "csave": "gzip",
+    "bz": "bz2",
+    "lzma": "lzma",
+    "lza": "lzma"
+}
+
+_DEFAULT_COMPRESSION_WRITE_MODES = {
+    "json": "w",
+    "gzip": "wt",
+    "bz2": "wt",
+    "lzma": "wt"
+}
+
+_DEFAULT_COMPRESSION_READ_MODES = {
+    "json": "r",
+    "gzip": "rt",
+    "bz2": "rt",
+    "lzma": "rt"
+}
+
+
+def get_compression_write_mode(compression: str) -> str:
+    """Return mode for opening file buffer for writing."""
+    return _DEFAULT_COMPRESSION_WRITE_MODES[compression]
+
+
+def get_compression_read_mode(compression: str) -> str:
+    """Return mode for opening file buffer for reading."""
+    return _DEFAULT_COMPRESSION_READ_MODES[compression]
+
+
+def infer_compression_from_filename(filename: str) -> str:
+    """Return the compression protocal inferred from given filename.
+    Parameters
+    ----------
+    filename: str
+        The filename for which to infer the compression protocol
+    """
+    return _DEFAULT_EXTENSION_MAP[filename.split(".")[-1]]
+
+
+def dump(obj: Any, path: str, compression_kwargs: Dict = None, json_kwargs: Dict = None):
+    """Dump the contents of an object to disk as json, to the supplied path, using the detected compression protocol.
+    
+    Parameters
+    ----------------
+    obj: any
+        The object that will be saved to disk
+    path: str
+        The path to the file to which to dump ``obj``
+    compression_kwargs:
+        keywords argument to pass to the compressed file opening protocol.
+    json_kwargs:
+        keywords argument to pass to the json file opening protocol.
+
+    Raises
+    ----------------
+    ValueError,
+        If given path is not a valid string.
+    """
+    if not isinstance(path, str):
+        raise ValueError("The given path is not a string.")
+    compression_kwargs = {} if compression_kwargs is None else compression_kwargs
+    json_kwargs = {} if json_kwargs is None else json_kwargs
+    compression = infer_compression_from_filename(path)
+    mode = get_compression_write_mode(compression)
+
+    directory = os.path.dirname(path)
+    if directory:
+        os.makedirs(directory, exist_ok=True)
+
+    if compression == "json":
```

Comment:
Maybe something like this could be used?
```python
if compression is None:
    compression = "gzip"

if compression is "json":
    open_function = open
else:
    compression_tool = importlib.import_module(compression)
    open_function = compression_tool.open

file = open_function(
    path,
    mode=mode,
    encoding="utf-8",
    **compression_kwargs,
)
```